Form Consumer now persists custom field value as donor and/or donation meta.

kjohnson opened the pull request:
<!-- Indicate the issue(s) resolved by this PR. -->

Resolves #5625

## Description

<!-- Summarize the related issue, explain HOW this PR solves the problem, and WHY you made the choices you made. -->

- Custom fields can be stored as donor meta with the fluent API method `storeAsDonorMeta()`.
- Custom fields can be stored as donation meta with the fluent API method `storeAsDonationMeta()`.

```php
add_action( 'give_fields_after_donation_levels', function( $fieldCollection, $formID ) {
	$fieldCollection->append(
		give_field( 'text', 'my-custom-field' )
			->label( 'My Custom Field' )
			->required()
			->storeAsDonorMeta()
	);
}, 10, 2 );
```

## Visuals

<!-- Include screenshots or video to better communicate your changes. -->

![image](https://user-images.githubusercontent.com/10858303/109199777-d8f4e200-776d-11eb-8240-2a87cc6d21be.png)

## Pre-review Checklist

<!-- Complete tasks prior to requesting a review. Add to this list, but do not remove the base items. -->

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@since` tags included in DocBlocks
-   [x] Changes logged to the `Unreleased` section of `CHANGELOG.md`
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

## Testing Instructions

<!-- Help others test your PR as efficiently as possible. -->

JasonTheAdams commented on src/Form/LegacyConsumer/Commands/SetupFieldPersistance.php:
```diff
@@ -21,7 +21,15 @@ function( $donationID, $donationData ) use ( $hook ) {
 							$value = wp_strip_all_tags( $_POST[ $field->getName() ], true );
-							give_update_payment_meta( $donationID, $field->getName(), $value );
+
+							if ( $field->shouldStoreAsDonationMeta() ) {
+								give_update_payment_meta( $donationID, $field->getName(), $value );
+							}
+
+							if ( $field->shouldStoreAsDonorMeta() ) {
+								$donorID = give_get_payment_meta( $donationID, '_give_payment_donor_id' );
+								Give()->donor_meta->update_meta( $donorID, $field->getName(), $value );
+							}
```
I would like to make this a bit more generic to make it easy for folks to provide their own storage methods:
```suggestion
							switch($field->storageType) {
								case Field::DONATION_STORAGE_TYPE:
									give_update_payment_meta( $donationID, $field->getName(), $value );
									break;
									
								case Field::DONOR_STORAGE_TYPE:
									$donorID = give_get_payment_meta( $donationID, '_give_payment_donor_id' );
									Give()->donor_meta->update_meta( $donorID, $field->getName(), $value );
									break;
								
								default:
									do_action('give_store_form_field', $field->storageType, $field, $value);
									do_action("give_store_form_field_{$field->storageType}", $field, $value);
							}
```
Something like that, anyway. What do you think?

kjohnson replied:
I don't think that generalizing it is necessary, as this is already a convenience added on top of the existing `give_insert_payment` action. The current approach provides a fluent, declarative path for storing as the most common donor or donation meta. If they wanted to handle it themselves then the hooks already exist for that.

kjohnson replied:
Also, I have separated the donor and donation meta flags as separate properties so that both can be used. I think it would make sense to still want the data collected by the donation to exist on the donation for record keeping while also storing the meta on the donor.

kjohnson replied:
Resolved in a call, see https://github.com/impress-org/givewp/pull/5652#issuecomment-788260103.

kjohnson commented:
## Call Summary

with @JasonTheAdams, @kjohnson

We discussed how custom field values should be stored and what options we should provide to developers writing custom code. This concluded with the following:

- Values should be stored as Donation Meta by default.
- Values can optionally be stored as Donor Meta instead of Donation Meta (either/or).
- Custom storage methods can be handled using the underlying `give_insert_payment` hook.

kjohnson commented:
@JasonTheAdams I addressed the points that we covered in our call, see https://github.com/impress-org/givewp/pull/5652/commits/9ef077d21bbd217fbf7fe2311ca271f3599417d6.

JasonTheAdams commented on src/Framework/FieldsAPI/FormField/FieldStoreAsMeta.php:
```diff
@@ -0,0 +1,32 @@
+	 * @unreleased
+	 * @var bool
+	 */
+	public $storeAsDonorMeta = false;
```
Let's make most properties `protected` as the default. I'd prefer folks to interact with fields declaratively.

JasonTheAdams replied:
I went ahead and did this so we can just merge it.

JasonTheAdams reviewed (approved):
Works great! Just one nitpicky thing, but I'm approving so go ahead and merge when done.
